syntax = "proto3";

package kvstore;

service KVStore {
    rpc Put(PutRequest) returns (PutResponse);
    rpc Get(GetRequest) returns (GetResponse);
    rpc Delete(DeleteRequest) returns (DeleteResponse);
    rpc Join(JoinRequest) returns (JoinResponse);
    rpc Leave(LeaveRequest) returns (LeaveResponse);
}

message PutRequest {
    string key = 1;
    string value = 2;
}
message PutResponse {
    bool success = 1;
    string error = 2;
}

message GetRequest {
    string key = 1;
}
message GetResponse {
    bool found = 1;
    string value = 2;
    string error = 3;
}

message DeleteRequest {
    string key = 1;
}
message DeleteResponse {
    bool success = 1;
    string error = 2;
}

message JoinRequest {
    string node_id = 1;
    string raft_endpoint = 2;
}
message JoinResponse {
    bool success = 1;
    string error = 2;
}

// Leaveリクエスト
message LeaveRequest {
    string node_id = 1;
}
message LeaveResponse {
    bool success = 1;
    string error = 2;
}